### PR TITLE
Task DVD-425

### PR DESCRIPTION
**Pull request by TheoWolf**

### CODE REVIEW

File: src/main/cpp/FfmpegJavaAvPlayback.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+void FfmpegJavaAvPlayback::set_audioSyncDelay(long lMillis){
+}
+
+long FfmpegJavaAvPlayback::get_audioSyncDelay() {
+	return 0;
+}
+
+// Has image data to display by checking the video frameque
+// different than has_image_data of the VideoState
+bool FfmpegJavaAvPlayback::has_image_data() {
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlayback.cpp]: Why is this needed?

File: src/main/cpp/FfmpegJavaAvPlayback.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+			return true;
+		}
+	}
+
+	return false;
+}
+
+// Has image data to display by checking the audio frameque
+// different than has_audio_data of the VideoState
+bool FfmpegJavaAvPlayback::has_audio_data() {
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlayback.cpp]: Why is this needed?

File: src/main/cpp/FfmpegJavaAvPlayback.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		if (pVideoState->get_pSampq()->nb_remaining() > 0) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+int FfmpegJavaAvPlayback::get_image_width() {
+	return pVideoState->get_video_st() ? pVideoState->get_image_width() : 0;
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlayback.cpp]: push the ternary logic that returns 0 into pVideoState->get_image_width()

File: src/main/cpp/FfmpegJavaAvPlayback.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+int FfmpegJavaAvPlayback::get_image_height() {
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlayback.cpp]: push the ternary logic that returns 0 into pVideoState->get_image_height()

File: src/main/cpp/FfmpegJavaAvPlayback.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	return pVideoState->get_video_st() ? pVideoState->get_image_height() : 0;
+}
+
+AudioFormat FfmpegJavaAvPlayback::get_audio_format() {
+	return AudioFormat();
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlayback.cpp]: Add TODO here

File: src/main/cpp/FfmpegJavaAvPlayback.cpp [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+PixelFormat FfmpegJavaAvPlayback::get_pixel_format() {
+	return PixelFormat();
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlayback.cpp]: Add TODO here

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+	});
+
+	pJavaPlayback->init_and_start_display_loop();
+
+	return ERROR_NONE;
+}
+
+void FfmpegJavaAvPlaybackPipline::Dispose() {
+	//TODO(Reda): re-visit destroy function for the java playback
+	//pJavaPlayback->destroy();
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: What's happening here?

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+		return ERROR_PLAYER_NULL;
+	}
+
+	*pfBalance = pJavaPlayback->get_balance();
+
+	return ERROR_NONE;
+}
+
+uint32_t FfmpegJavaAvPlaybackPipline::SetAudioSyncDelay(long lMillis) {
+	// TODO(fraudies): Implement this
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Remove comment

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+		return ERROR_PLAYER_NULL;
+	}
+
+	pJavaPlayback->set_audioSyncDelay(lMillis);
+
+	return ERROR_NONE; // no error
+}
+
+uint32_t FfmpegJavaAvPlaybackPipline::GetAudioSyncDelay(long* plMillis) {
+	// TODO(fraudies): Implement this
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Remove this comment

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+	*plMillis = pJavaPlayback->get_audioSyncDelay();
+
+	return ERROR_NONE; // no error
+}
+
+uint32_t FfmpegJavaAvPlaybackPipline::HasAudioData(bool* bAudioData) const {
+	if (pJavaPlayback == nullptr) {
+		return ERROR_PLAYER_NULL;
+	}
+	// TODO(fraudies): Implement this
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Remove this comment

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+	*bAudioData = pJavaPlayback->has_audio_data();
+
+	return ERROR_NONE;
+}
+
+uint32_t FfmpegJavaAvPlaybackPipline::HasImageData(bool* bImageData) const {
+	if (pJavaPlayback == nullptr) {
+		return ERROR_PLAYER_NULL;
+	}
+	// TODO(fraudies): Implement this
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Remove this comment

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+	*bImageData = pJavaPlayback->has_image_data();
+
+	return ERROR_NONE;
+}
+
+uint32_t FfmpegJavaAvPlaybackPipline::GetImageWidth(int* width) const {
+	if (pJavaPlayback == nullptr) {
+		return ERROR_PLAYER_NULL;
+	}
+	// TODO(fraudies): Implement this
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Remove this comment

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+	*width = pJavaPlayback->get_image_width();
+
+	return ERROR_NONE;
+}
+
+uint32_t FfmpegJavaAvPlaybackPipline::GetImageHeight(int* iHeight) const {
+	if (pJavaPlayback == nullptr) {
+		return ERROR_PLAYER_NULL;
+	}
+	// TODO(fraudies): Implement this
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Remove this comment

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+	*iHeight = pJavaPlayback->get_image_height();
+
+	return ERROR_NONE;
+}
+
+uint32_t FfmpegJavaAvPlaybackPipline::GetAudioFormat(AudioFormat* pAudioFormat) const {
+	if (pJavaPlayback == nullptr) {
+		return ERROR_PLAYER_NULL;
+	}
+	// TODO(fraudies): Implement this
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Remove this comment

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+	}
+	// TODO(fraudies): Implement the pixel format for the ouput
+	*pPixelFormat = pJavaPlayback->get_pixel_format();
+
+	return ERROR_NONE;
+}
+
+uint32_t FfmpegJavaAvPlaybackPipline::GetImageBuffer(uint8_t** ppImageBuffer) {
+	if (pJavaPlayback == nullptr) {
+		return ERROR_PLAYER_NULL;
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Remove this comment

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 16]
```diff
@@ -0,0 +1,368 @@
...
+
+uint32_t FfmpegJavaAvPlaybackPipline::GetAudioBuffer(uint8_t** ppAudioBuffer) {
+	if (pJavaPlayback == nullptr) {
+		return ERROR_PLAYER_NULL;
+	}
+	// TODO(fraudies): Implement the audio buffer data
+	return ERROR_NONE;
+}
+
+void FfmpegJavaAvPlaybackPipline::UpdatePlayerState(PlayerState newState) {
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Add TODO: Push this logic into a super-class that is common the SDL / Java implementation

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp [hunk 17]
```diff
@@ -0,0 +1,368 @@
...
+	case Error:
+		break;
+	}
+
+	// The same thread can acquire the same lock several times
+	SetPlayerState(newPlayerState, bSilent);
+}
+
+
+void FfmpegJavaAvPlaybackPipline::SetPlayerState(PlayerState newPlayerState, bool bSilent) {
```
Comment (fraudies) [inline src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Add TODO: Push this into a super-class that is common to the SDL/Java implementation

File: src/main/cpp/FfmpegPipelineFactory.cpp [hunk 18]
```diff
@@ -8,15 +8,14 @@ FfmpegPipelineFactory::~FfmpegPipelineFactory()
 
 uint32_t FfmpegPipelineFactory::CreatePlayerPipeline(CPipelineOptions *pOptions, CPipeline **ppPipeline) {
 
-	if (NULL == pOptions->GetStreamData())
-		return ERROR_BASE_PIPELINE;
+	//if (NULL == pOptions->GetStreamData())
+	//	return ERROR_BASE_PIPELINE;
 	
 	*ppPipeline = NULL;
 	if (pOptions->GetStreamData())
```
Comment (fraudies) [inline src/main/cpp/FfmpegPipelineFactory.cpp]: Do you have the logic inverted?

I thought if getStreamData is set then we would like to use the playback through Java.

File: src/main/cpp/FfmpegJavaAvPlayback.cpp (hunk 1 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlayback.cpp]: This return true if there frames to display in the FrameQue, compared to one in the VideoState class that will return true if there is a video stream available, might be useful to rename one of them

Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlayback.cpp]: Removed

File: src/main/cpp/FfmpegJavaAvPlayback.cpp (hunk 2 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlayback.cpp]: Removed

File: src/main/cpp/FfmpegJavaAvPlayback.cpp (hunk 3 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlayback.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlayback.cpp (hunk 4 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlayback.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlayback.cpp (hunk 5 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlayback.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlayback.cpp (hunk 6 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlayback.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 7 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Fixed

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 8 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 9 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 10 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 11 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 12 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 13 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 14 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 15 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Done

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 16 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Added.

File: src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp (hunk 17 above)
Comment (TheoWolf) [reply src/main/cpp/FfmpegJavaAvPlaybackPipline.cpp]: Added.
